File: Examples/UIExplorer/android/app/src/main/java/com/facebook/react/uiapp/UpdatePropertiesExampleView.java
/**
 * The examples provided by Facebook are for non-commercial testing and
 * evaluation purposes only.
 *
 * Facebook reserves all rights not expressly granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL
 * FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.react.uiapp;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;

import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactRootView;
import com.facebook.react.uimanager.PixelUtil;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;

import static com.facebook.react.common.ApplicationHolder.getApplication;

public class UpdatePropertiesExampleView extends LinearLayout {

  ReactRootView rootView;
  Button button;
  boolean beige;

  public UpdatePropertiesExampleView(Context context) {
    super(context);

    setOrientation(VERTICAL);

    beige = true;

    ReactApplication reactApplication = (ReactApplication)getApplication();
    ReactNativeHost reactNativeHost = reactApplication.getReactNativeHost();
    ReactInstanceManager reactInstanceManager = reactNativeHost.getReactInstanceManager();

    Bundle launchOptions = new Bundle();
    launchOptions.putString("color", "beige");

    rootView = new ReactRootView(context);
    rootView.startReactApplication(reactInstanceManager, "SetPropertiesExampleApp", launchOptions);

    int spaceHeight = (int)PixelUtil.toPixelFromDIP(20);
    LayoutParams rootViewLayoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1);
    rootViewLayoutParams.setMargins(0, 0, 0, spaceHeight);

    addView(rootView, rootViewLayoutParams);

    button = new Button(context);
    button.setText("Native Button");
    button.setTextColor(Color.WHITE);
    button.setBackgroundColor(Color.GRAY);
    button.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View v) {
        UpdatePropertiesExampleView.this.changeColor();
      }
    });

    int buttonHeight = (int)PixelUtil.toPixelFromDIP(40);
    LayoutParams buttonLayoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, buttonHeight);

    addView(button, buttonLayoutParams);
  }

  @Override
  public void onDetachedFromWindow() {
    super.onDetachedFromWindow();

    rootView.unmountReactApplication();
    removeView(rootView);
    rootView = null;
  }

  public void changeColor() {
    beige = !beige;

    Bundle launchOptions = new Bundle();
    launchOptions.putString("color", beige ? "beige" : "purple");

    rootView.updateLaunchOptions(launchOptions);
  }

  public static class Manager extends ViewGroupManager<UpdatePropertiesExampleView> {

    @Override
    public String getName() {
      return "UpdatePropertiesExampleView";
    }

    @Override
    public void addView(UpdatePropertiesExampleView parent, View child, int index) {
      //Do nothing
    }

    @Override
    protected UpdatePropertiesExampleView createViewInstance(ThemedReactContext reactContext) {
      return new UpdatePropertiesExampleView(reactContext);
    }
  }
}

File: Examples/UIExplorer/UIExplorer/NativeExampleViews/UpdatePropertiesExampleView.m
/**
 * The examples provided by Facebook are for non-commercial testing and
 * evaluation purposes only.
 *
 * Facebook reserves all rights not expressly granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL
 * FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#import "UpdatePropertiesExampleView.h"

#import <React/RCTRootView.h>
#import <React/RCTViewManager.h>

#import "AppDelegate.h"

@interface UpdatePropertiesExampleViewManager : RCTViewManager

@end

@implementation UpdatePropertiesExampleViewManager

RCT_EXPORT_MODULE();

- (UIView *)view
{
  return [UpdatePropertiesExampleView new];
}

@end

@implementation UpdatePropertiesExampleView
{
  RCTRootView *_rootView;
  UIButton *_button;
  BOOL _beige;
}

- (instancetype)initWithFrame:(CGRect)frame
{
  self = [super initWithFrame:frame];
  if (self) {
    _beige = YES;

    AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];

    _rootView = [[RCTRootView alloc] initWithBridge:appDelegate.bridge
                                         moduleName:@"SetPropertiesExampleApp"
                                  initialProperties:@{@"color":@"beige"}];

    _button = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [_button setTitle:@"Native Button" forState:UIControlStateNormal];
    [_button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [_button setBackgroundColor:[UIColor grayColor]];

    [_button addTarget:self
                action:@selector(changeColor)
      forControlEvents:UIControlEventTouchUpInside];

    [self addSubview:_button];
    [self addSubview:_rootView];
  }
  return self;
}

- (void)layoutSubviews
{
  float spaceHeight = 20;
  float buttonHeight = 40;
  float rootViewWidth = self.bounds.size.width;
  float rootViewHeight = self.bounds.size.height - spaceHeight - buttonHeight;

  [_rootView setFrame:CGRectMake(0, 0, rootViewWidth, rootViewHeight)];
  [_button setFrame:CGRectMake(0, rootViewHeight + spaceHeight, rootViewWidth, buttonHeight)];
}

- (void)changeColor
{
  _beige = !_beige;
  [_rootView setAppProperties:@{@"color":_beige ? @"beige" : @"purple"}];
}

- (NSArray<UIView<RCTComponent> *> *)reactSubviews
{
  // this is to avoid unregistering our RCTRootView when the component is removed from RN hierarchy
  (void)[super reactSubviews];
  return @[];
}

@end
